app/rpc/namespaces/eth/tx_pool.go

```diff
@@ -0,0 +1,306 @@
+		return err
+	}
+	defer itr.Close()
+	for ; itr.Valid(); itr.Next() {
+		key := string(itr.Key())
+		txBytes := itr.Value()
+
+		tmp := strings.Split(key, "|")
+		if len(tmp) != 2 {
+			continue
+		}
+		address := common.HexToAddress(tmp[0])
+		txNonce, err := strconv.Atoi(tmp[1])
+		if err != nil {
+			return err
+		}
+
+		tx := new(evmtypes.MsgEthereumTx)
+		if err = rlp.DecodeBytes(txBytes, tx); err != nil {
+			return err
+		}
+		if int(tx.Data.AccountNonce) != txNonce {
+			return fmt.Errorf("nonce[%d] in key is not equal to nonce[%d] in value", tx.Data.AccountNonce, txNonce)
+		}
+
+		pCurrentNonce, err := api.GetTransactionCount(address, rpctypes.PendingBlockNumber)
+		if err != nil {
+			return err
+		}
+		currentNonce := int(*pCurrentNonce)
+		if txNonce < currentNonce {
+			continue
+		}
+
+		if err = pool.insertTx(address, tx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func broadcastTxByTxPool(api *PublicEthereumAPI, tx *evmtypes.MsgEthereumTx, txBytes []byte) (common.Hash, error) {
+	// Get sender address
+	chainIDEpoch, err := ethermint.ParseChainID(api.clientCtx.ChainID)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	from, err := tx.VerifySig(chainIDEpoch)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	api.txPool.mu.Lock()
+	if err = api.txPool.CacheAndBroadcastTx(api, from, tx); err != nil {
+		api.logger.Error("eth_sendRawTransaction txPool err:", err.Error())
+		api.txPool.mu.Unlock()
+		return common.Hash{}, err
+	}
+	api.txPool.mu.Unlock()
+
+	return common.HexToHash(strings.ToUpper(hex.EncodeToString(tmhash.Sum(txBytes)))), nil
+}
+
+func (pool *TxPool) CacheAndBroadcastTx(api *PublicEthereumAPI, address common.Address, tx *evmtypes.MsgEthereumTx) error {
+	// get currentNonce
+	pCurrentNonce, err := api.GetTransactionCount(address, rpctypes.PendingBlockNumber)
+	if err != nil {
+		return err
+	}
+	currentNonce := uint64(*pCurrentNonce)
+
+	if tx.Data.AccountNonce < currentNonce {
+		return fmt.Errorf("AccountNonce of tx is less than currentNonce in memPool: AccountNonce[%d], currentNonce[%d]", tx.Data.AccountNonce, currentNonce)
+	}
+
+	if tx.Data.AccountNonce > currentNonce+viper.GetUint64(TxPoolSliceMaxLen) {
+		return fmt.Errorf("AccountNonce of tx is bigger than txPool capacity, please try later: AccountNonce[%d]", tx.Data.AccountNonce)
+	}
+
+	if err = pool.insertTx(address, tx); err != nil {
+		return err
+	}
+
+	// update DB
+	if err = pool.writeTxInDB(address, tx); err != nil {
+		return err
+	}
+
+	return pool.continueBroadcast(api, currentNonce, address)
+}
+
+func (pool *TxPool) update(index int, address common.Address, tx *evmtypes.MsgEthereumTx) error {
+	txsLen := len(pool.addressTxsPool[address])
+	if index >= txsLen {
+		pool.addressTxsPool[address] = append(pool.addressTxsPool[address], tx)
+	} else {
+		tmpTx := make([]*evmtypes.MsgEthereumTx, len(pool.addressTxsPool[address][index:]))
+		copy(tmpTx, pool.addressTxsPool[address][index:])
+
+		pool.addressTxsPool[address] =
```

Comment:
Using list instead of slice to cache tx can reduce memory copy when inserting new tx or deleting tx in the header, thereby reducing performance loss.

Reply:
fix it later